### PR TITLE
[Frontend] Added Deregister Functionality

### PR DESCRIPTION
**Pull request by Suryansh5545**
Added Backend api and frontend button to deregister participant.
Edge Cases:-
When participant team already have a submission
when participant team isn't in challenge.

https://github.com/Cloud-CV/EvalAI/assets/34577232/f0f59aba-1385-4e7e-a259-52636ac6380f

### CODE REVIEW

File: apps/challenges/views.py [hunk 1]
```diff
@@ -316,6 +316,46 @@ def challenge_detail(request, challenge_host_team_pk, challenge_pk):
...
+        all_challenge_phases = ChallengePhase.objects.filter(challenge=challenge)
+        if (all_challenge_phases.count() > 0):
+            for challenge_phase in all_challenge_phases:
+                submission_exist = Submission.objects.filter(participant_team=participant_team, challenge_phase=challenge_phase).exists()
+                if submission_exist:
+                    break
+        else:
+            submission_exist = False
+        if submission_exist:
+            response_data = {"error": "You cannot deregister as you already have a submission in the challenge"}
```
Comment (gchhablani) [inline apps/challenges/views.py:348]: `"Participant teams which have made submissions to a challenge cannot be deregistered."`

File: apps/challenges/views.py [hunk 2]
```diff
@@ -316,6 +316,46 @@ def challenge_detail(request, challenge_host_team_pk, challenge_pk):
...
+            return Response(response_data, status=status.HTTP_406_NOT_ACCEPTABLE)
+        else:
+            challenge.participant_teams.remove(participant_team)
+            response_data = {"success": "Successfully deregistered!"}
+            return Response(response_data, status=status.HTTP_200_OK)
+    else:
+        response_data = {"error": "You are not registered in this challenge"}
```
Comment (gchhablani) [inline apps/challenges/views.py:355]: `"Your participant team is not registered for this challenge."`

File: frontend/src/views/web/challenge/submission.html [hunk 3]
```diff
@@ -27,6 +27,12 @@ <h5 class="w-300">Submission Guidelines</h5>
         </div>
     </div>
 </section>
+<section ng-if="!challenge.isChallengeHost" class="ev-sm-container ev-view challenge-container">
+    <div class="ev-md-container ev-card-panel ev-z-depth-5">
+        Click here to deregister from this challenge:
+        <button class="btn ev-btn-dark waves-effect waves-dark grad-btn grad-btn-dark fs-16" ng-click="challenge.deregisterdialog($event)"> Deregister</button>
```
Comment (gchhablani) [inline frontend/src/views/web/challenge/submission.html:33]: Why the space before `Deregister`?

File: frontend/src/views/web/challenge/submission.html [hunk 4]
```diff
@@ -178,7 +184,6 @@ <h5 class="w-300">
             <div class="col s12">
                 <h5 class="w-300">Make Submission</h5>
             </div>
-
```
Comment (gchhablani) [inline frontend/src/views/web/challenge/submission.html:181]: Why remove this line?

File: frontend/src/views/web/challenge/submission.html [hunk 5]
```diff
@@ -509,6 +514,6 @@ <h5 class="w-300">Make Submission</h5>
 <!-- Show error when the challenge is not active -->
 <section ng-if="!challenge.isActive" class="ev-sm-container ev-view challenge-container">
     <div class="ev-md-container ev-card-panel ev-z-depth-5 ev-challenge-view">
-        <p class="fs-16 w-300">Sorry, the challenge is not active.</p>
```
Comment (gchhablani) [inline frontend/src/views/web/challenge/submission.html:512]: Why remove this paragraph?

File: tests/unit/challenges/test_views.py [hunk 6]
```diff
@@ -310,6 +310,39 @@ def test_remove_participant_team(self):
...
+    def test_deregister_participant_team(self):
+        self.url = reverse_lazy(
+            "challenges:deregister_participant_team_from_challenge",
+            kwargs={"challenge_pk": self.challenge.pk},
+        )
+
+        response = self.client.post(self.url, {})
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data["success"], "Successfully deregistered!")
+
```
Comment (gchhablani) [inline tests/unit/challenges/test_views.py:334]: Is it not possible to add a test for the case when there are existing submissions?

**Review by gchhablani — COMMENTED** [review]
Added some comments.

File: frontend/src/views/web/challenge/submission.html (hunk 3 above)
Comment (Suryansh5545) [reply frontend/src/views/web/challenge/submission.html:33]: sorry it was a typo

File: frontend/src/views/web/challenge/submission.html (hunk 4 above)
Comment (Suryansh5545) [reply frontend/src/views/web/challenge/submission.html:181]: typo

File: frontend/src/views/web/challenge/submission.html (hunk 5 above)
Comment (Suryansh5545) [reply frontend/src/views/web/challenge/submission.html:512]: yea this line should stay i will fix it

File: tests/unit/challenges/test_views.py (hunk 6 above)
Comment (Suryansh5545) [reply tests/unit/challenges/test_views.py:394]: I checked and yes we can do it too , i will add it

**Pull request merged**